File: apps/go/simulator/biology.go
package simulator

import (
	"fmt"
	"math"
	"math/rand"
	"sort"
	"sync/atomic"
)

// EpigeneticMark represents an epigenetic modification at a specific genomic position.
type EpigeneticMark struct {
	Position        int
	MarkType        string // "methylation", "acetylation", "phosphorylation"
	Active          bool
	GenerationAdded int
}

// Gene represents a gene: a segment of DNA that encodes a protein.
type Gene struct {
	Name            string
	Sequence        []string // List of bases
	StartPos        int
	EndPos          int
	ExpressionLevel float64 // 0.0 = silenced, 1.0 = fully active
	EpigeneticMarks []EpigeneticMark
	Essential       bool
}

// NewGene creates a new gene.
func NewGene(name string, sequence []string, startPos, endPos int, essential bool) *Gene {
	g := &Gene{
		Name:            name,
		Sequence:        sequence,
		StartPos:        startPos,
		EndPos:          endPos,
		ExpressionLevel: 1.0,
		EpigeneticMarks: make([]EpigeneticMark, 0),
		Essential:       essential,
	}
	return g
}

// Length returns the gene length.
func (g *Gene) Length() int {
	return len(g.Sequence)
}

// IsSilenced returns true if the gene is heavily methylated.
func (g *Gene) IsSilenced() bool {
	methylCount := 0
	for _, m := range g.EpigeneticMarks {
		if m.MarkType == "methylation" && m.Active {
			methylCount++
		}
	}
	return float64(methylCount) > float64(g.Length())*0.3
}

// Methylate adds a methylation mark.
func (g *Gene) Methylate(position, generation int) {
	if position >= 0 && position < g.Length() {
		g.EpigeneticMarks = append(g.EpigeneticMarks, EpigeneticMark{
			Position:        position,
			MarkType:        "methylation",
			Active:          true,
			GenerationAdded: generation,
		})
		g.updateExpression()
	}
}

// Demethylate removes a methylation mark at a position.
func (g *Gene) Demethylate(position int) {
	filtered := g.EpigeneticMarks[:0]
	for _, m := range g.EpigeneticMarks {
		if !(m.Position == position && m.MarkType == "methylation") {
			filtered = append(filtered, m)
		}
	}
	g.EpigeneticMarks = filtered
	g.updateExpression()
}

// Acetylate adds histone acetylation.
func (g *Gene) Acetylate(position, generation int) {
	g.EpigeneticMarks = append(g.EpigeneticMarks, EpigeneticMark{
		Position:        position,
		MarkType:        "acetylation",
		Active:          true,
		GenerationAdded: generation,
	})
	g.updateExpression()
}

func (g *Gene) updateExpression() {
	methyl := 0
	acetyl := 0
	for _, m := range g.EpigeneticMarks {
		if m.MarkType == "methylation" && m.Active {
			methyl++
		}
		if m.MarkType == "acetylation" && m.Active {
			acetyl++
		}
	}
	length := g.Length()
	if length == 0 {
		length = 1
	}
	suppression := math.Min(1.0, float64(methyl)/float64(length)*3)
	activation := math.Min(1.0, float64(acetyl)/float64(length)*5)
	g.ExpressionLevel = math.Max(0.0, math.Min(1.0, 1.0-suppression+activation))
}

// Transcribe converts DNA to mRNA (T -> U).
func (g *Gene) Transcribe() []string {
	if g.IsSilenced() {
		return nil
	}
	rna := make([]string, len(g.Sequence))
	for i, base := range g.Sequence {
		if base == "T" {
			rna[i] = "U"
		} else {
			rna[i] = base
		}
	}
	return rna
}

// Mutate applies random point mutations. Returns mutation count.
func (g *Gene) Mutate(rate float64, rng *rand.Rand) int {
	mutations := 0
	for i := 0; i < g.Length(); i++ {
		if rng.Float64() < rate {
			old := g.Sequence[i]
			choices := make([]string, 0, 3)
			for _, b := range NucleotideBases {
				if b != old {
					choices = append(choices, b)
				}
			}
			g.Sequence[i] = choices[rng.Intn(len(choices))]
			mutations++
		}
	}
	return mutations
}

// DNAStrand represents a double-stranded DNA molecule.
type DNAStrand struct {
	Sequence      []string
	Genes         []*Gene
	Generation    int
	MutationCount int
}

// complement maps for DNA base pairing.
var complement = map[string]string{
	"A": "T", "T": "A", "G": "C", "C": "G",
}

// Length returns the DNA strand length.
func (d *DNAStrand) Length() int {
	return len(d.Sequence)
}

// GCContent returns the GC content fraction.
func (d *DNAStrand) GCContent() float64 {
	if len(d.Sequence) == 0 {
		return 0.0
	}
	gc := 0
	for _, b := range d.Sequence {
		if b == "G" || b == "C" {
			gc++
		}
	}
	return float64(gc) / float64(len(d.Sequence))
}

// RandomStrand generates a random DNA strand with genes.
func RandomStrand(length, numGenes int, rng *rand.Rand) *DNAStrand {
	sequence := make([]string, length)
	for i := range sequence {
		sequence[i] = NucleotideBases[rng.Intn(len(NucleotideBases))]
	}
	strand := &DNAStrand{Sequence: sequence, Genes: make([]*Gene, 0, numGenes)}

	geneLen := length / (numGenes + 1)
	for i := 0; i < numGenes; i++ {
		quarter := geneLen / 4
		if quarter < 1 {
			quarter = 1
		}
		start := i*geneLen + rng.Intn(quarter)
		end := start + geneLen/2
		if end > length {
			end = length
		}

		geneSeq := make([]string, end-start)
		copy(geneSeq, sequence[start:end])

		gene := NewGene(
			fmt.Sprintf("gene_%d", i),
			geneSeq,
			start,
			end,
			i == 0, // first gene is essential
		)
		strand.Genes = append(strand.Genes, gene)
	}

	return strand
}

// Replicate performs semi-conservative replication with possible errors.
func (d *DNAStrand) Replicate(rng *rand.Rand) *DNAStrand {
	newSeq := make([]string, len(d.Sequence))
	copy(newSeq, d.Sequence)

	newGenes := make([]*Gene, 0, len(d.Genes))
	for _, gene := range d.Genes {
		geneSeq := make([]string, len(gene.Sequence))
		copy(geneSeq, gene.Sequence)

		// Epigenetic marks can be partially inherited
		marks := make([]EpigeneticMark, 0)
		for _, m := range gene.EpigeneticMarks {
			if rng.Float64() < 0.7 { // Some marks lost in replication
				marks = append(marks, EpigeneticMark{
					Position:        m.Position,
					MarkType:        m.MarkType,
					Active:          m.Active && rng.Float64() < 0.8,
					GenerationAdded: m.GenerationAdded,
				})
			}
		}

		newGene := NewGene(gene.Name, geneSeq, gene.StartPos, gene.EndPos, gene.Essential)
		newGene.EpigeneticMarks = marks
		newGene.updateExpression()
		newGenes = append(newGenes, newGene)
	}

	return &DNAStrand{
		Sequence:   newSeq,
		Genes:      newGenes,
		Generation: d.Generation + 1,
	}
}

// ApplyMutations applies environmental mutations.
func (d *DNAStrand) ApplyMutations(uvIntensity, cosmicRayFlux float64, rng *rand.Rand) int {
	totalMutations := 0
	rate := UVMutationRate*uvIntensity + CosmicRayMutationRate*cosmicRayFlux

	for _, gene := range d.Genes {
		m := gene.Mutate(rate, rng)
		totalMutations += m
	}

	// Also mutate non-genic regions
	for i := 0; i < d.Length(); i++ {
		if rng.Float64() < rate {
			old := d.Sequence[i]
			choices := make([]string, 0, 3)
			for _, b := range NucleotideBases {
				if b != old {
					choices = append(choices, b)
				}
			}
			d.Sequence[i] = choices[rng.Intn(len(choices))]
			totalMutations++
		}
	}

	d.MutationCount += totalMutations
	return totalMutations
}

// ApplyEpigeneticChanges applies environmental epigenetic modifications.
func (d *DNAStrand) ApplyEpigeneticChanges(temperature float64, generation int, rng *rand.Rand) {
	for _, gene := range d.Genes {
		// Methylation
		if rng.Float64() < MethylationProbability {
			gLen := gene.Length()
			if gLen == 0 {
				gLen = 1
			}
			pos := rng.Intn(gLen)
			gene.Methylate(pos, generation)
		}

		// Demethylation
		if rng.Float64() < DemethylationProbability {
			var methyls []EpigeneticMark
			for _, m := range gene.EpigeneticMarks {
				if m.MarkType == "methylation" {
					methyls = append(methyls, m)
				}
			}
			if len(methyls) > 0 {
				mark := methyls[rng.Intn(len(methyls))]
				gene.Demethylate(mark.Position)
			}
		}

		// Histone acetylation (temperature-dependent)
		thermalFactor := math.Min(2.0, temperature/300.0)
		if rng.Float64() < HistoneAcetylationProb*thermalFactor {
			gLen := gene.Length()
			if gLen == 0 {
				gLen = 1
			}
			pos := rng.Intn(gLen)
			gene.Acetylate(pos, generation)
		}

		// Histone deacetylation
		if rng.Float64() < HistoneDeacetylationProb {
			var acetyls []*EpigeneticMark
			for i := range gene.EpigeneticMarks {
				if gene.EpigeneticMarks[i].MarkType == "acetylation" {
					acetyls = append(acetyls, &gene.EpigeneticMarks[i])
				}
			}
			if len(acetyls) > 0 {
				mark := acetyls[rng.Intn(len(acetyls))]
				mark.Active = false
				gene.updateExpression()
			}
		}
	}
}

// TranslateMRNA translates mRNA to protein (amino acid sequence).
func TranslateMRNA(mrna []string) []string {
	var protein []string
	i := 0
	started := false

	for i+2 < len(mrna) {
		codon := mrna[i] + mrna[i+1] + mrna[i+2]
		aa, ok := CodonTable[codon]

		if aa == "Met" && !started {
			started = true
			protein = append(protein, aa)
		} else if started {
			if aa == "STOP" {
				break
			} else if ok {
				protein = append(protein, aa)
			}
		}
		i += 3
	}

	return protein
}

// Protein represents a chain of amino acids.
type Protein struct {
	AminoAcids []string
	Name       string
	Function   string // "enzyme", "structural", "signaling"
	Folded     bool
	Active     bool
}

// NewProtein creates a new protein.
func NewProtein(aminoAcids []string, name, function string) *Protein {
	return &Protein{
		AminoAcids: aminoAcids,
		Name:       name,
		Function:   function,
		Active:     true,
	}
}

// Length returns the protein length.
func (p *Protein) Length() int {
	return len(p.AminoAcids)
}

// Fold performs simplified protein folding.
func (p *Protein) Fold(rng *rand.Rand) bool {
	if p.Length() < 3 {
		p.Folded = false
		return false
	}
	foldProb := math.Min(0.9, 0.5+0.1*math.Log(float64(p.Length()+1)))
	p.Folded = rng.Float64() < foldProb
	return p.Folded
}

// cellIDGen is the global cell ID counter.
var cellIDGen atomic.Int64

// Cell represents a cell with DNA, RNA, and proteins.
type Cell struct {
	DNA        *DNAStrand
	Proteins   []*Protein
	Fitness    float64
	Alive      bool
	Generation int
	Energy     float64
	ID         int64
	rng        *rand.Rand
}

// NewCell creates a new cell.
func NewCell(dna *DNAStrand, generation int, rng *rand.Rand) *Cell {
	return &Cell{
		DNA:        dna,
		Proteins:   make([]*Protein, 0),
		Fitness:    1.0,
		Alive:      true,
		Generation: generation,
		Energy:     100.0,
		ID:         cellIDGen.Add(1),
		rng:        rng,
	}
}

// TranscribeAndTranslate implements the central dogma: DNA -> mRNA -> Protein.
func (c *Cell) TranscribeAndTranslate() []*Protein {
	var newProteins []*Protein
	functions := []string{"enzyme", "structural", "signaling"}

	for _, gene := range c.DNA.Genes {
		if gene.ExpressionLevel < 0.1 {
			continue
		}

		mrna := gene.Transcribe()
		if len(mrna) == 0 {
			continue
		}

		aaSeq := TranslateMRNA(mrna)
		if len(aaSeq) == 0 {
			continue
		}

		if c.rng.Float64() > gene.ExpressionLevel {
			continue
		}

		protein := NewProtein(aaSeq, "protein_"+gene.Name, functions[c.rng.Intn(len(functions))])
		protein.Fold(c.rng)
		newProteins = append(newProteins, protein)
		c.Proteins = append(c.Proteins, protein)
	}

	return newProteins
}

// Metabolize performs basic metabolism.
func (c *Cell) Metabolize(environmentEnergy float64) {
	enzymeCount := 0
	for _, p := range c.Proteins {
		if p.Function == "enzyme" && p.Folded && p.Active {
			enzymeCount++
		}
	}
	efficiency := 0.3 + 0.15*float64(enzymeCount)
	c.Energy += environmentEnergy * efficiency
	c.Energy -= 3.0 // Basal metabolic cost
	if c.Energy > 200.0 {
		c.Energy = 200.0
	}
	if c.Energy <= 0 {
		c.Alive = false
	}
}

// Divide performs cell division with DNA replication.
func (c *Cell) Divide() *Cell {
	if !c.Alive || c.Energy < 50.0 {
		return nil
	}

	newDNA := c.DNA.Replicate(c.rng)
	c.Energy /= 2

	daughter := NewCell(newDNA, c.Generation+1, c.rng)
	daughter.Energy = c.Energy
	daughter.TranscribeAndTranslate()

	return daughter
}

// ComputeFitness computes cell fitness based on functional proteins and DNA integrity.
func (c *Cell) ComputeFitness() float64 {
	if !c.Alive {
		c.Fitness = 0.0
		return 0.0
	}

	// Essential genes must be active
	essentialActive := true
	for _, g := range c.DNA.Genes {
		if g.Essential && g.IsSilenced() {
			essentialActive = false
			break
		}
	}
	if !essentialActive {
		c.Fitness = 0.1
		return 0.1
	}

	// Fitness from proteins
	functionalProteins := 0
	for _, p := range c.Proteins {
		if p.Folded && p.Active {
			functionalProteins++
		}
	}
	numGenes := len(c.DNA.Genes)
	if numGenes == 0 {
		numGenes = 1
	}
	proteinFitness := math.Min(1.0, float64(functionalProteins)/float64(numGenes))

	// Fitness from energy
	energyFitness := math.Min(1.0, c.Energy/100.0)

	// GC content near 0.5 is optimal
	gcFitness := 1.0 - math.Abs(c.DNA.GCContent()-0.5)*2

	c.Fitness = proteinFitness*0.4 + energyFitness*0.3 + gcFitness*0.3
	return c.Fitness
}

// Biosphere is a collection of cells with population dynamics.
type Biosphere struct {
	Cells      []*Cell
	Generation int
	TotalBorn  int
	TotalDied  int
	DNALength  int
	rng        *rand.Rand
}

// NewBiosphere creates a new biosphere.
func NewBiosphere(initialCells, dnaLength int, rng *rand.Rand) *Biosphere {
	bio := &Biosphere{
		Cells:     make([]*Cell, 0, 128),
		DNALength: dnaLength,
		rng:       rng,
	}

	for i := 0; i < initialCells; i++ {
		dna := RandomStrand(dnaLength, 3, rng)
		cell := NewCell(dna, 0, rng)
		cell.TranscribeAndTranslate()
		bio.Cells = append(bio.Cells, cell)
		bio.TotalBorn++
	}

	return bio
}

// Step performs one generation step.
func (bio *Biosphere) Step(environmentEnergy, uvIntensity, cosmicRayFlux, temperature float64) {
	bio.Generation++

	// Metabolize
	for _, cell := range bio.Cells {
		cell.Metabolize(environmentEnergy)
	}

	// Apply mutations
	for _, cell := range bio.Cells {
		if cell.Alive {
			cell.DNA.ApplyMutations(uvIntensity, cosmicRayFlux, bio.rng)
			cell.DNA.ApplyEpigeneticChanges(temperature, bio.Generation, bio.rng)
		}
	}

	// Transcribe/translate (clear old proteins to prevent unbounded growth)
	for _, cell := range bio.Cells {
		if cell.Alive {
			cell.Proteins = cell.Proteins[:0]
			cell.TranscribeAndTranslate()
		}
	}

	// Compute fitness
	for _, cell := range bio.Cells {
		cell.ComputeFitness()
	}

	// Selection and reproduction
	var aliveCells []*Cell
	for _, c := range bio.Cells {
		if c.Alive {
			aliveCells = append(aliveCells, c)
		}
	}

	if len(aliveCells) > 0 {
		sort.Slice(aliveCells, func(i, j int) bool {
			return aliveCells[i].Fitness > aliveCells[j].Fitness
		})
		cutoff := len(aliveCells) / 2
		if cutoff < 1 {
			cutoff = 1
		}
		var newCells []*Cell
		for _, cell := range aliveCells[:cutoff] {
			daughter := cell.Divide()
			if daughter != nil {
				newCells = append(newCells, daughter)
				bio.TotalBorn++
			}
		}
		bio.Cells = append(bio.Cells, newCells...)
	}

	// Remove dead cells
	dead := 0
	for _, c := range bio.Cells {
		if !c.Alive {
			dead++
		}
	}
	bio.TotalDied += dead

	alive := make([]*Cell, 0, len(bio.Cells))
	for _, c := range bio.Cells {
		if c.Alive {
			alive = append(alive, c)
		}
	}
	bio.Cells = alive

	// Population cap
	if len(bio.Cells) > 100 {
		sort.Slice(bio.Cells, func(i, j int) bool {
			return bio.Cells[i].Fitness > bio.Cells[j].Fitness
		})
		bio.TotalDied += len(bio.Cells) - 100
		bio.Cells = bio.Cells[:100]
	}
}

// AverageFitness returns the average fitness of all cells.
func (bio *Biosphere) AverageFitness() float64 {
	if len(bio.Cells) == 0 {
		return 0.0
	}
	total := 0.0
	for _, c := range bio.Cells {
		total += c.Fitness
	}
	return total / float64(len(bio.Cells))
}

// AverageGCContent returns the average GC content.
func (bio *Biosphere) AverageGCContent() float64 {
	if len(bio.Cells) == 0 {
		return 0.0
	}
	total := 0.0
	for _, c := range bio.Cells {
		total += c.DNA.GCContent()
	}
	return total / float64(len(bio.Cells))
}

// TotalMutations returns the total mutations across all cells.
func (bio *Biosphere) TotalMutations() int {
	total := 0
	for _, c := range bio.Cells {
		total += c.DNA.MutationCount
	}
	return total
}

File: apps/rust/src/simulator/biology.rs
//! Biological systems simulation.
//!
//! Models DNA replication, mutation, protein synthesis, lifeforms with traits,
//! natural selection, and population dynamics. Emerges from the chemical
//! substrate once nucleotides and amino acids are available.

use rand::Rng;
use std::collections::HashMap;

use super::constants::*;
use super::chemistry::ChemicalSystem;

// ---------------------------------------------------------------------------
// DNA / Genome
// ---------------------------------------------------------------------------

/// A strand of DNA represented as a sequence of nucleotide bases.
#[derive(Debug, Clone)]
pub struct Genome {
    pub bases: Vec<char>,
    pub methylation: Vec<bool>,
    pub generation: u64,
}

impl Genome {
    /// Create a random genome of the given length.
    pub fn random(length: usize) -> Self {
        let mut rng = rand::thread_rng();
        let bases: Vec<char> = (0..length)
            .map(|_| NUCLEOTIDE_BASES[rng.gen_range(0..4)])
            .collect();
        let methylation = vec![false; length];
        Self {
            bases,
            methylation,
            generation: 0,
        }
    }

    /// Transcribe DNA to RNA (T -> U).
    pub fn transcribe(&self) -> Vec<char> {
        self.bases
            .iter()
            .map(|&b| if b == 'T' { 'U' } else { b })
            .collect()
    }

    /// Translate RNA into a protein (sequence of amino acid names).
    pub fn translate(&self) -> Vec<&'static str> {
        let rna = self.transcribe();
        let mut protein = Vec::new();
        let mut i = 0;

        // Scan for start codon AUG
        while i + 2 < rna.len() {
            let codon: String = rna[i..i + 3].iter().collect();
            if codon == "AUG" {
                break;
            }
            i += 1;
        }

        // Translate codons
        while i + 2 < rna.len() {
            let codon: String = rna[i..i + 3].iter().collect();
            match codon_to_amino_acid(&codon) {
                Some("STOP") => break,
                Some(aa) => protein.push(aa),
                None => {}
            }
            i += 3;
        }

        protein
    }

    /// Mutate the genome with a given mutation rate per base.
    pub fn mutate(&mut self, mutation_rate: f64) {
        let mut rng = rand::thread_rng();
        for i in 0..self.bases.len() {
            // Methylated bases have reduced mutation rate
            let effective_rate = if self.methylation[i] {
                mutation_rate * 0.1
            } else {
                mutation_rate
            };

            if rng.gen::<f64>() < effective_rate {
                // Point mutation
                let new_base = NUCLEOTIDE_BASES[rng.gen_range(0..4)];
                self.bases[i] = new_base;
            }
        }

        // Epigenetic changes: methylation / demethylation
        for i in 0..self.methylation.len() {
            if !self.methylation[i] && rng.gen::<f64>() < METHYLATION_PROBABILITY {
                self.methylation[i] = true;
            } else if self.methylation[i] && rng.gen::<f64>() < DEMETHYLATION_PROBABILITY {
                self.methylation[i] = false;
            }
        }
    }

    /// Replicate with possible errors.
    pub fn replicate(&self, error_rate: f64) -> Self {
        let mut child = self.clone();
        child.generation = self.generation + 1;
        child.mutate(error_rate);
        child
    }

    /// Count occurrences of each base.
    pub fn base_composition(&self) -> HashMap<char, usize> {
        let mut counts = HashMap::new();
        for &b in &self.bases {
            *counts.entry(b).or_insert(0) += 1;
        }
        counts
    }

    /// GC content as a fraction.
    pub fn gc_content(&self) -> f64 {
        let gc = self.bases.iter().filter(|&&b| b == 'G' || b == 'C').count();
        gc as f64 / self.bases.len().max(1) as f64
    }

    /// Fraction of methylated bases.
    pub fn methylation_level(&self) -> f64 {
        let m = self.methylation.iter().filter(|&&x| x).count();
        m as f64 / self.methylation.len().max(1) as f64
    }

    pub fn len(&self) -> usize {
        self.bases.len()
    }
}

// ---------------------------------------------------------------------------
// Trait / Phenotype
// ---------------------------------------------------------------------------

/// A phenotypic trait derived from the genome.
#[derive(Debug, Clone)]
pub struct Trait {
    pub name: String,
    pub value: f64,
}

/// Decode traits from a genome. Uses simple hashing of subsequences.
pub fn decode_traits(genome: &Genome) -> Vec<Trait> {
    let mut traits = Vec::new();
    let len = genome.len();
    if len < 12 {
        return traits;
    }

    // Heat resistance: encoded by first 12 bases
    let heat_val: f64 = genome.bases[..12]
        .iter()
        .enumerate()
        .map(|(i, &b)| {
            let v = match b {
                'A' => 0.0,
                'T' => 1.0,
                'G' => 2.0,
                'C' => 3.0,
                _ => 0.0,
            };
            v / 4.0_f64.powi(i as i32 + 1)
        })
        .sum();
    traits.push(Trait {
        name: "heat_resistance".to_string(),
        value: heat_val.min(1.0),
    });

    // Metabolic efficiency: next 12 bases
    if len >= 24 {
        let met_val: f64 = genome.bases[12..24]
            .iter()
            .enumerate()
            .map(|(i, &b)| {
                let v = match b {
                    'A' => 0.0,
                    'T' => 1.0,
                    'G' => 2.0,
                    'C' => 3.0,
                    _ => 0.0,
                };
                v / 4.0_f64.powi(i as i32 + 1)
            })
            .sum();
        traits.push(Trait {
            name: "metabolic_efficiency".to_string(),
            value: met_val.min(1.0),
        });
    }

    // Reproduction rate: next 12 bases
    if len >= 36 {
        let rep_val: f64 = genome.bases[24..36]
            .iter()
            .enumerate()
            .map(|(i, &b)| {
                let v = match b {
                    'A' => 0.0,
                    'T' => 1.0,
                    'G' => 2.0,
                    'C' => 3.0,
                    _ => 0.0,
                };
                v / 4.0_f64.powi(i as i32 + 1)
            })
            .sum();
        traits.push(Trait {
            name: "reproduction_rate".to_string(),
            value: rep_val.min(1.0),
        });
    }

    traits
}

// ---------------------------------------------------------------------------
// Lifeform
// ---------------------------------------------------------------------------

/// A single lifeform with a genome and fitness.
#[derive(Debug, Clone)]
pub struct Lifeform {
    pub genome: Genome,
    pub traits: Vec<Trait>,
    pub fitness: f64,
    pub energy: f64,
    pub age: u64,
    pub alive: bool,
    pub species_id: u64,
}

static SPECIES_COUNTER: std::sync::atomic::AtomicU64 =
    std::sync::atomic::AtomicU64::new(0);

fn next_species_id() -> u64 {
    SPECIES_COUNTER.fetch_add(1, std::sync::atomic::Ordering::Relaxed) + 1
}

impl Lifeform {
    /// Create a new lifeform with a random genome.
    pub fn new(genome_length: usize) -> Self {
        let genome = Genome::random(genome_length);
        let traits = decode_traits(&genome);
        Self {
            genome,
            traits,
            fitness: 0.5,
            energy: 100.0,
            age: 0,
            alive: true,
            species_id: next_species_id(),
        }
    }

    /// Create a lifeform from an existing genome.
    pub fn from_genome(genome: Genome) -> Self {
        let traits = decode_traits(&genome);
        Self {
            genome,
            traits,
            fitness: 0.5,
            energy: 100.0,
            age: 0,
            alive: true,
            species_id: next_species_id(),
        }
    }

    /// Get a trait value by name, defaulting to 0.5.
    pub fn trait_value(&self, name: &str) -> f64 {
        self.traits
            .iter()
            .find(|t| t.name == name)
            .map(|t| t.value)
            .unwrap_or(0.5)
    }

    /// Evaluate fitness against environmental pressures.
    pub fn evaluate_fitness(&mut self, temperature: f64, resources: f64) {
        let heat_res = self.trait_value("heat_resistance");
        let metabolism = self.trait_value("metabolic_efficiency");

        // Temperature fitness: penalty for extreme temps
        let optimal_temp = 288.0 + heat_res * 100.0; // 288K - 388K range
        let temp_diff = (temperature - optimal_temp).abs();
        let temp_fitness = (-temp_diff / 200.0).exp();

        // Resource fitness
        let resource_fitness = (resources * metabolism).min(1.0);

        self.fitness = (temp_fitness * 0.5 + resource_fitness * 0.5).max(0.0).min(1.0);
    }

    /// Age and consume energy. Returns false if the lifeform dies.
    pub fn tick(&mut self) -> bool {
        if !self.alive {
            return false;
        }

        self.age += 1;
        let metabolism = self.trait_value("metabolic_efficiency");
        let energy_cost = 1.0 + (1.0 - metabolism) * 2.0;
        self.energy -= energy_cost;

        if self.energy <= 0.0 || self.fitness < 0.01 {
            self.alive = false;
            return false;
        }

        true
    }

    /// Reproduce with mutation.
    pub fn reproduce(&self, mutation_rate: f64) -> Option<Lifeform> {
        if !self.alive || self.energy < 50.0 {
            return None;
        }

        let child_genome = self.genome.replicate(mutation_rate);
        let mut child = Lifeform::from_genome(child_genome);
        child.energy = self.energy * 0.4;
        Some(child)
    }

    /// Protein the genome encodes.
    pub fn protein(&self) -> Vec<&'static str> {
        self.genome.translate()
    }
}

// ---------------------------------------------------------------------------
// BiologicalSystem
// ---------------------------------------------------------------------------

/// Manages a population of lifeforms with selection.
pub struct BiologicalSystem {
    pub population: Vec<Lifeform>,
    pub generation: u64,
    pub total_born: u64,
    pub total_died: u64,
    pub total_mutations: u64,
    pub max_population: usize,
    pub base_mutation_rate: f64,
}

impl BiologicalSystem {
    pub fn new() -> Self {
        Self {
            population: Vec::new(),
            generation: 0,
            total_born: 0,
            total_died: 0,
            total_mutations: 0,
            max_population: 200,
            base_mutation_rate: 0.001,
        }
    }

    /// Seed the population with initial lifeforms from available chemistry.
    pub fn abiogenesis(&mut self, chem: &ChemicalSystem) -> usize {
        let available_nucleotides = chem.nucleotide_count;
        let available_amino_acids = chem.amino_acid_count;

        if available_nucleotides < 10 || available_amino_acids < 5 {
            return 0;
        }

        let mut rng = rand::thread_rng();
        // Probability scales with available precursors
        let prob = (available_nucleotides as f64 * available_amino_acids as f64) / 1e6;
        if rng.gen::<f64>() > prob && !self.population.is_empty() {
            return 0;
        }

        // Create a few primitive lifeforms
        let count = rng.gen_range(1..=3).min(5 - self.population.len().min(4));
        let mut created = 0;
        for _ in 0..count {
            if self.population.len() >= self.max_population {
                break;
            }
            let genome_len = rng.gen_range(30..90);
            let lf = Lifeform::new(genome_len);
            self.population.push(lf);
            self.total_born += 1;
            created += 1;
        }
        created
    }

    /// Run one generation: evaluate fitness, select, reproduce, cull.
    pub fn tick(&mut self, temperature: f64, resources: f64, radiation: f64) {
        if self.population.is_empty() {
            return;
        }

        self.generation += 1;

        // Mutation rate increases with radiation
        let mutation_rate = self.base_mutation_rate + radiation * UV_MUTATION_RATE;

        // Evaluate fitness and age
        for lf in &mut self.population {
            lf.evaluate_fitness(temperature, resources);
            lf.tick();
            // Feed: energy proportional to fitness and resources
            if lf.alive {
                lf.energy += lf.fitness * resources * 5.0;
            }
        }

        // Reproduction
        let mut offspring = Vec::new();
        let mut rng = rand::thread_rng();
        for lf in &mut self.population {
            if !lf.alive {
                continue;
            }
            let rep_rate = lf.trait_value("reproduction_rate");
            if rng.gen::<f64>() < rep_rate * lf.fitness * 0.3 {
                if let Some(mut child) = lf.reproduce(mutation_rate) {
                    lf.energy -= 30.0;
                    child.evaluate_fitness(temperature, resources);
                    offspring.push(child);
                    self.total_born += 1;
                    self.total_mutations += 1;
                }
            }
        }
        self.population.extend(offspring);

        // Remove dead
        let before = self.population.len();
        self.population.retain(|lf| lf.alive);
        self.total_died += (before - self.population.len()) as u64;

        // Population cap: keep the fittest
        if self.population.len() > self.max_population {
            self.population
                .sort_by(|a, b| b.fitness.partial_cmp(&a.fitness).unwrap());
            let died = self.population.len() - self.max_population;
            self.population.truncate(self.max_population);
            self.total_died += died as u64;
        }
    }

    /// Average fitness of the population.
    pub fn average_fitness(&self) -> f64 {
        if self.population.is_empty() {
            return 0.0;
        }
        let sum: f64 = self.population.iter().map(|lf| lf.fitness).sum();
        sum / self.population.len() as f64
    }

    /// Average genome length.
    pub fn average_genome_length(&self) -> f64 {
        if self.population.is_empty() {
            return 0.0;
        }
        let sum: f64 = self.population.iter().map(|lf| lf.genome.len() as f64).sum();
        sum / self.population.len() as f64
    }

    /// Number of unique species.
    pub fn species_count(&self) -> usize {
        let ids: std::collections::HashSet<u64> =
            self.population.iter().map(|lf| lf.species_id).collect();
        ids.len()
    }

    pub fn to_compact(&self) -> String {
        format!(
            "BIO[pop={} gen={} born={} died={} mut={} fit={:.3} glen={:.0} spp={}]",
            self.population.len(),
            self.generation,
            self.total_born,
            self.total_died,
            self.total_mutations,
            self.average_fitness(),
            self.average_genome_length(),
            self.species_count(),
        )
    }
}
